### PR TITLE
New Triple Valve Features

### PR DESCRIPTION
**Pull request by SteelFill**
My first enhancement for OR, this adds multiple new features to the air brake code to allow for simulating some features of more modern brake valve types (eg: ABD, ABDW), and also refactors/reorganizes some brake system behaviors to accommodate the new features better.

[(Start of) Discussion on ElvasTower](http://www.elvastower.com/forums/index.php?/topic/34527-wishes-for-improvement-of-braking-systems/page__view__findpost__p__298101)
[Trello Card](https://trello.com/c/el72p3b7/569-new-features-for-american-triple-valves)

- Refactor of triple valve state updates
- Smoothing added to final stages of application/release to hopefully prevent triple valve from rapidly toggling states
- Changes to aux/emerg res charging behavior to better accommodate graduated release systems
- Quick release: Emerg res air can be piped into the brake pipe during release to speed up release. Enable with **`ORTSEmergencyResQuickRelease ( 1 )`** [default is 0]
- Uniform charging: Aux res charging can be slowed down when the brake pipe pressure is substantially higher than the aux res pressure to allow faster releases. Set the difference between BP and AUX to activate uniform charging with **`ORTSUniformChargingThreshold ( 3psi )`** [default is 3 psi], and the factor by which the rate of charging is reduced with **`ORTSUniformChargingRatio ( 2 )`** [default is 0, disabling the feature] {setting of 2 halves the rate, 0.5 would double the rate}.
- Quick service: When changing from release to apply, the brake valve will vent some air from the brake pipe to speed up initial applications until the brake cylinder pressure has reached a certain level. Set the target brake cylinder feed line pressure using **`ORTSQuickServiceLimit ( 10psi )`** [default is 0, disabling the feature], set the rate of brake pipe venting with **`ORTSQuickServiceVentRate ( 2.5psi/s )`** [default is 0] or leave it at 0 psi/s to simulate systems which don't vent the brake pipe, and optionally allow for brake cylinder feed pressure to increase faster than usual using **`ORTSQuickServiceApplicationRate ( 15psi/s )`** [default is 0].
- Accelerated application: During applications, the brake valve will measure the rate of reduction of the brake pipe and will vent air from the brake pipe in proportion to the rate of reduction to dramatically speed up service applications. Set the proportionality constant of air vented to measured reduction using **`ORTSAcceleratedApplicationFactor ( 0.5 )`** [default is 0, disabling the feature] but beware that this is an open loop control system, and setting ratios too large can cause too much air to be vented (I found factors above 1 start to risk the brake pipe emptying itself). Additionally, set a limit on the venting rate of accelerated application with **`ORTSAcceleratedApplicationMaxVentRate ( 1psi/s )`** [default is 5 psi].
- Sensitivity to initial reductions: Not all brake valves are built equal, and some will not change to 'apply' when in 'release' until the brake pipe pressure drops by a few psi. The pressure difference required to trip the valve from release to apply can be set with **`ORTSInitialApplicationThreshold ( 2psi )`** [default is 1 psi of direct release, 0.15 bar for graduated release].
- Brake cylinder springs: The return spring inside each brake cylinder will result in some very low brake cylinder pressures producing no force at the brake pads. If desired, the minimum brake cylinder pressure for any brake force to result can be set with **`ORTSCylinderSpringPressure ( 3psi )`** [default is 0, resulting in force with any brake cylinder pressure].
- Separate pressure limits in service and emergency: Some brake systems will have different brake cylinder pressure limits for service applications than for emergency applications. A limit on the output of the triple valve for service applications can be set with **`ORTSMaxServiceCylinderPressure ( 72psi )`** [default is 0, resulting in this limit being ignored], while `ORTSMaxTripleValveCylinderPressure` acts as the pressure limit in emergency.
- Accelerated emergency release: On cars with emergency reservoirs, release after an emergency application is sped up by pushing air from the brake cylinder and auxiliary reservoir into the brake pipe. This occurs when the brake pipe pressure is restored above 20 psi. This feature has no tokens as I haven't seen any evidence that the behavior varies substantially between different types of brake valve. It will be disabled entirely if `ORTSEmergencyDumpValveTimer` is set to 0.

### CODE REVIEW

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/AirSinglePipe.cs [hunk 1]
```diff
@@ -256,7 +294,19 @@ public override void Save(BinaryWriter outf)
...
+            outf.Write(UniformChargingRatio);
+            outf.Write(QuickServiceActive);
+            outf.Write(QuickServiceLimitPSI);
+            outf.Write(QuickServiceApplicationRatePSIpS);
+            outf.Write(QuickServiceVentRatePSIpS);
+            outf.Write(AcceleratedApplicationRatio);
+            outf.Write(AcceleratedApplicationLimitPSIpS);
+            outf.Write(InitialApplicationThresholdPSI);
+            outf.Write(BrakeCylinderSpringPressurePSI);
+            outf.Write(ServiceMaxCylPressurePSI);
```
Comment (cesarBLG) [inline Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/AirSinglePipe.cs:308]: You don't need to save things that are constant (e.g. application rates, volumes, etc.). Only things that may change during the simulation need to be changed (e.g. QuickServiceActive or current cyl pressure).

**Review by cesarBLG — CHANGES_REQUESTED** [review]
Thanks for the improvements! They look OK to me, unfortunately they will conflict with my PR. I'll try to find reviewers for my PRs, and as soon as they are merged I'll add this PR to the Unstable releases

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/AirSinglePipe.cs (hunk 1 above)
Comment (SteelFill) [reply Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/AirSinglePipe.cs:308]: Good to know, I wasn't entirely sure how saving interacted with initialization-I'll get rid of those things that don't need to be re initialized.

**Comment by cesarBLG** [comment]
My two PRs have been approved and merged (thanks Peter). You can now merge the master branch into your branch and solve conflicts. Basically, the conflicting point will be the creation of two different Initialize() methods. One is intended for initial configuration of brakes (e.g. to configure parameters to a default value if they are not defined), and the other one is to reset brakes with shift+/

After conflicts are solved, your changes will be available in the Unstable release, as I've tagged this PR as "for-unstable"

**Comment by SteelFill** [comment]
Got the conflicts with the relay valve code sorted, and I did a scan to make sure I'm using AutoCylPressurePSI instead of CylPressurePSI which should mean my code will function as expected with relay valves (all pressures are in terms of the brake cylinder feed line, not the actual brake cylinder pressure). That said, I am not at home so I have not had an opportunity to actually do a build and test that theory.

**Comment by cesarBLG** [comment]
Could you please add a Trello card for the feature, and link it here? Also, a link to the relevant forum posts is required. (see https://github.com/openrails/openrails/blob/master/Docs/Contributing.md for further information). Once the Trello card is approved, I can review and merge this PR.

**Comment by cjakeman** [comment]
> Could you please add a Trello card for the feature, and link it here? 

[Trello card](https://trello.com/c/el72p3b7/569-new-features-for-american-triple-valves)

**Comment by SteelFill** [comment]
Thanks for approving the Trello card Chris!

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/AirSinglePipe.cs [hunk 2]
```diff
@@ -486,17 +636,42 @@ public override void Update(float elapsedClockSeconds)
...
                     if ((Car as MSTSWagon).EmergencyReservoirPresent)
                     {
-                        dp = elapsedClockSeconds * MaxApplicationRatePSIpS;
-                        if (EmergResPressurePSI - dp < AuxResPressurePSI + dp * EmergAuxVolumeRatio)
-                            dp = (EmergResPressurePSI - AuxResPressurePSI) / (1 + EmergAuxVolumeRatio);
-                        EmergResPressurePSI -= dp;
-                        AuxResPressurePSI += dp * EmergAuxVolumeRatio;
+                        if (EmergencyDumpValveTimerS != 0 && EmergencyDumpStartTime == null && BrakeLine1PressurePSI > AcceleratedEmergencyReleaseThresholdPSI)
+                        {
+                            // Accelerated emergency release: Aux res and BP air are routed into the brake pipe once the emergency application is complete, speeds up emergency release
```
Comment (cesarBLG) [inline Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/AirSinglePipe.cs:650]:
```suggestion
                            // Accelerated emergency release: Aux res and BC air are routed into the brake pipe once the emergency application is complete, speeds up emergency release
```

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/AirSinglePipe.cs [hunk 3]
```diff
@@ -472,12 +589,45 @@ public override void Update(float elapsedClockSeconds)
             // triple valve is set to charge the brake cylinder
             if ((TripleValveState == ValveState.Apply || TripleValveState == ValveState.Emergency) && !Car.WheelBrakeSlideProtectionActive)
             {
-                float dp = elapsedClockSeconds * MaxApplicationRatePSIpS;
+                float dp = 0;
+                float dpPipe = 0;
+                float BrakePipeChange = Math.Max(prevBrakePipePressurePSI - BrakeLine1PressurePSI, 0);
```
Comment (cesarBLG) [inline Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/AirSinglePipe.cs:594]:
```suggestion
                float brakePipeChange = Math.Max(prevBrakePipePressurePSI - BrakeLine1PressurePSI, 0);
```
Please use lowercase names for local variables

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/AirSinglePipe.cs [hunk 4]
```diff
@@ -569,30 +718,90 @@ public override void Update(float elapsedClockSeconds)
...
+                        float dp = elapsedClockSeconds * MaxAuxilaryChargingRatePSIpS;
+                        if (AuxResPressurePSI + dp > EmergResPressurePSI - dp / EmergAuxVolumeRatio)
+                            dp = (EmergResPressurePSI - AuxResPressurePSI) * EmergAuxVolumeRatio / (1 + EmergAuxVolumeRatio);
+                        if (BrakeLine1PressurePSI < AuxResPressurePSI + dp)
+                            dp = (BrakeLine1PressurePSI - AuxResPressurePSI);
+                        AuxResPressurePSI += dp;
+                        EmergResPressurePSI -= dp / EmergAuxVolumeRatio;
+                    }
+                }
+                if (AuxResPressurePSI > EmergResPressurePSI && (valveType == MSTSWagon.BrakeValveType.Distributor ? true : TripleValveState == ValveState.Release))
```
Comment (cesarBLG) [inline Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/AirSinglePipe.cs:761]:
```suggestion
                if (AuxResPressurePSI > EmergResPressurePSI && (valveType == MSTSWagon.BrakeValveType.Distributor || TripleValveState == ValveState.Release))
```

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/AirSinglePipe.cs [hunk 5]
```diff
@@ -569,30 +718,90 @@ public override void Update(float elapsedClockSeconds)
...
+                    if (AuxResPressurePSI + dpAux > BrakeLine2PressurePSI - dpAux * AuxBrakeLineVolumeRatio)
+                        dpAux = (BrakeLine2PressurePSI - AuxResPressurePSI) / (1 + AuxBrakeLineVolumeRatio);
+                    AuxResPressurePSI += dpAux;
+                    BrakeLine2PressurePSI -= dpAux * AuxBrakeLineVolumeRatio;
+                }
+                
+            }
+            else // Charge from brake pipe
+            {
+                if (AuxResPressurePSI < BrakeLine1PressurePSI && (valveType == MSTSWagon.BrakeValveType.Distributor ? true : TripleValveState == ValveState.Release) && !BleedOffValveOpen)
```
Comment (cesarBLG) [inline Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/AirSinglePipe.cs:787]:
```suggestion
                if (AuxResPressurePSI < BrakeLine1PressurePSI && (valveType == MSTSWagon.BrakeValveType.Distributor || TripleValveState == ValveState.Release) && !BleedOffValveOpen)
```

**Review by cesarBLG — COMMENTED** [review]
I have just a couple of suggestions, nothing really important. Thanks for all the new features!

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/AirSinglePipe.cs (hunk 2 above)
Comment (SteelFill) [reply Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/AirSinglePipe.cs:650]: Derp on my part, good catch.

**Review by cesarBLG — APPROVED** [review]
Thanks, everything looks OK to me.

**Pull request merged**